### PR TITLE
Draft PR for code review: Pytorch implementation

### PR DESCRIPTION
**Pull request by Christianfoley**
PR for code review and commenting. See issue #175.

### CODE REVIEW

File: micro_dl/cli/torch_inference_script.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    args = parser.parse_args()
+    return args
+
+def check_save_folder(inf_cfig, prep_cfig):
+    '''
+    Helper method to ensure that save folder exists.
+    If no save folder specified in inference_config, force saving in data
+    directory with dynamic name and timestamp.
+    
+    :param pd.dataframe inference_config: inference config file (not) containing save_folder_name
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_inference_script.py:58]: docstring params don't match those in function definition. Personally I prefer the params in the docstring since they're easier to read and understand.

File: micro_dl/cli/torch_inference_script.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    :param pd.dataframe preprocess_config: preprocessing config file containing input_dir
+    '''
+    
+    if 'save_folder_name' not in inf_cfig:
+        assert 'input_dir' in prep_cfig, 'Error in autosaving: \'input_dir\'' \
+            'unspecified in preprocess config'
+        now = str(datetime.datetime.now()).replace(' ', '_').replace(':','_').replace('-','_')[:-10]
+        save_dir = os.path.join(prep_cfig['input_dir'], f'../prediction_{now}')
+        
+        if os.path.exists(save_dir):
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_inference_script.py:68]: If the path exists, create it?

File: micro_dl/cli/torch_inference_script.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import yaml
+import datetime
+import torch
+import micro_dl.inference.image_inference as image_inf
+import micro_dl.torch_unet.utils.inference as torch_inference_utils
+import micro_dl.utils.train_utils as train_utils
+
+import argparse
+
+def read_config(config_path):
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_inference_script.py:11]: There's already a read config function in microDL so please use that: https://github.com/mehta-lab/microDL/blob/master/micro_dl/utils/aux_utils.py#L47

File: micro_dl/cli/torch_inference_script.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import os
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_inference_script.py:1]: Tiny nitpicky detail, but it's nice to sort imports alphabetically, external imports first then those from the repo. It helps you see quicker what's there.

File: micro_dl/cli/torch_inference_script.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    
+    #if no save_folder_name specified, automatically incur saving in data folder
+    check_save_folder(inference_config, preprocess_config)
+
+    #instantiate and prep TorchPredictor interfacing object
+    torch_predictor = torch_inference_utils.TorchPredictor(network_config = network_config, device = device)
+    torch_predictor.load_model_torch()
+        
+    #instantiate ImagePredictor object and run inference
+    image_predictor = image_inf.ImagePredictor(train_config, inference_config, torch_predictor,
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_inference_script.py:99]: If you can't fit all function args in one line, it's easer to read if you put each on a new line.

File: micro_dl/cli/torch_train_script.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import yaml
+import argparse
+
+import micro_dl.torch_unet.utils.training as train
+
+
+def read_config(config_path):
```
Comment (jennyfolkesson) [inline micro_dl/cli/torch_train_script.py:7]: Please use existing config reader to avoid code redundancy.

File: micro_dl/cli/train_script.py [hunk 7]
```diff
@@ -149,6 +149,7 @@ def create_datasets(df_meta,
 
 
 def create_network(network_config, gpu_id):
+    raise DeprecationWarning('Tensorflow networks are no longer supported as of 2.0.0')
```
Comment (jennyfolkesson) [inline micro_dl/cli/train_script.py:152]: 🎉

File: micro_dl/config_inference_dataset.yml [hunk 8]
```diff
@@ -0,0 +1,69 @@
+dataset:
+  augmentation: false
+  data_dir: /hpc/projects/comp_micro/projects/virtualstaining/torch_microDL/data/2022_09_07_Soorya_KidneyRickettsia/data/
+  input_channels:
+  - 3
+  mask_channels:
+  - 4
+  model_task: regression
+  normalize: false
+  random_seed: 42
```
Comment (jennyfolkesson) [inline micro_dl/config_inference_dataset.yml:10]: You probably only want to set a fixed random seed when you're testing, right?

File: micro_dl/config_inference_dataset.yml [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  split_by_column: pos_idx
+  split_ratio:
+    test: 0.05
+    train: 0.05
+    val: 0.9
```
Comment (jennyfolkesson) [inline micro_dl/config_inference_dataset.yml:15]: lots of validation going on here 😄

File: micro_dl/config_inference_dataset.yml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    type: relu
+  batch_norm: true
+  block_sequence: conv-activation-bn
+  class: UNetStackTo2D
+  data_format: channels_first
+  depth: 5
+  dropout: 0.2
+  filter_size: 3
+  final_activation: linear
+  height: 2048
```
Comment (jennyfolkesson) [inline micro_dl/config_inference_dataset.yml:30]: maybe move width to be near height?

File: micro_dl/config_inference_dataset.yml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  num_convs_per_block: 2
+  num_filters_per_block:
+  - 16
+  - 32
+  - 64
+  - 128
+  - 256
+  num_input_channels: 1
```
Comment (jennyfolkesson) [inline micro_dl/config_inference_dataset.yml:38]: Since we already have specified which input and output channels we use, it seems a bit redundant to again specify how many of each there are, it's basically just len(input_channels) etc.
I know this is not your design but a remnant of bad design from way back when, but still it would be nice to remove these.

**Review by jennyfolkesson — COMMENTED** [review]
Just scratching the surface here. Great work, I'll keep adding comments as I work my way through this branch.

File: .vscode/launch.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply .vscode/launch.json:2]: I hadn't even noticed this was part of the repo. I think vscode automatically set this up. Good catch!

File: micro_dl/config_inference_dataset.yml (hunk 10 above)
Comment (Christianfoley) [reply micro_dl/config_inference_dataset.yml:30]: Yeah... These config files have some wonky stuff in them left over from the master branch. Soorya tackled a lot of this in the documentation pr, so I will pull those changes.

File: micro_dl/cli/torch_inference_script.py (hunk 2 above)
Comment (Christianfoley) [reply micro_dl/cli/torch_inference_script.py:59]: Good catch! This actually explains why my predictions sometimes were not saving.

File: micro_dl/config/_torch_config.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    optimizer: Literal['adam','sgd']
+    loss: Literal['mse','l1','cossim','cel']
+    testing_stride: int
+    save_model_stride: int
+    save_dir: str
+    mask: bool
+    mask_type: str
+    device: Union[str, int]
+    
+    @validator('save_dir')
```
Comment (jennyfolkesson) [inline micro_dl/config/_torch_config.py:50]: The valid_path and positive functions are identical to those in the model. 
I'm still learning pydantic, but to avoid lots of boiler plate, could you use the allow_reuse arg in validator?

File: micro_dl/config/_torch_config.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+               'save_model_string',
+               check_fields=False)
+    def positive(cls, n):
+        if not n > 0:
+            raise ValueError(f'{validator.__name__} must be greater than 0.')
+        return n
+    
+    @validator('device')
+    def valid_device(cls, v):
+        if isinstance(v, str):
```
Comment (jennyfolkesson) [inline micro_dl/config/_torch_config.py:70]: Add docstring

File: micro_dl/config/_torch_config.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+            assert v == 'cpu' or v == 'gpu', "Device must be 'cpu','gpu' (for"\
+                "autoselection), or a gpu index"
+        elif isinstance(v, int):
+            if v+1 > device_count():
```
Comment (jennyfolkesson) [inline micro_dl/config/_torch_config.py:74]: First isinstance check you raise an assertion error, here you raise a value error, perhaps raise same exception type?

File: micro_dl/config/_torch_config.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+                raise ValueError(f"gpu index {v} not in available gpus "\
+                    f"{range(device_count())}")
+        return v
```
Comment (jennyfolkesson) [inline micro_dl/config/_torch_config.py:77]: v is not modified, is it?

File: micro_dl/config/_torch_config.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    train_config_path : str
+    inference_config_path : str
+    
+    model: Union[_TorchModel, None]
+    training: Union[_TorchTraining, None]
+    
+    @validator('preprocess_config_path',
+               'train_config_path',
+               'inference_config_path')
+    def valid_path(cls, path):
```
Comment (jennyfolkesson) [inline micro_dl/config/_torch_config.py:94]: More boiler plate, see previous comment.

**Review by jennyfolkesson — COMMENTED** [review]
Chipping away on this PR with a few more comments... :-)

File: micro_dl/inference/image_inference.py [hunk 18]
```diff
@@ -722,11 +785,12 @@ def predict_2d(self, chan_slice_meta):
                         self.crop_shape,
                         self.image_format,
                     )
+                
+                #pass to network for prediction
```
Comment (jennyfolkesson) [inline micro_dl/inference/image_inference.py:789]: Some more nitpicking: for proper pep you need a space between your pound symbol and your comment.

File: micro_dl/inference/image_inference.py [hunk 19]
```diff
@@ -287,6 +309,7 @@ def _get_split_ids(self, data_split='test'):
         return split_col, inference_ids
 
     def _assign_3d_inference(self):
+        raise DeprecationWarning('3d inference no longer supported in 2.0.0')
```
Comment (jennyfolkesson) [inline micro_dl/inference/image_inference.py:312]: Since this is now deprecated, you can just delete all the 3D inference functions and tests and just keep the warning.

File: micro_dl/inference/image_inference.py [hunk 20]
```diff
@@ -784,7 +849,7 @@ def predict_2d(self, chan_slice_meta):
         target_stack = np.concatenate(target_stack, axis=0)
         # Stack images and transpose (metrics assumes cyxz format)
         if self.image_format == 'zyx':
-            if self.input_depth > 1:
+            if self.input_depth > 0:
```
Comment (jennyfolkesson) [inline micro_dl/inference/image_inference.py:852]: It looks like the default input_depth is 1. Is it ever <1? If not, do you even need the if statement?

File: micro_dl/config/_torch_config.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    architecture: Literal['2.5D', '2D']
+    in_channels: int
+    out_channels: int
+    residual: bool
+    task: Literal['reg','seg']
+    model_dir: str
+    
+    @validator('model_dir')
+    def valid_path(cls, path):
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/config/_torch_config.py:20]: I'm actually not sure how to avoid the boiler plate repetition in pydantic. I'm very much a beginner user myself. I had originally written only single functions but I couldn't get pydantic to pull the input values correctly, so this is the hacky fix I settled for. I think more time needs to be spent on this code; let's push this back to the pydantic branch PR when it's ready to merge.

File: micro_dl/inference/image_inference.py [hunk 22]
```diff
@@ -133,8 +155,8 @@ def __init__(self,
         self.name_format = 'ctzp'
         if 'name_format' in images_dict:
             self.name_format = images_dict['name_format']
+            
         # Create image subdirectory to write predicted images
-        self.pred_dir = os.path.join(self.model_dir, self.save_folder_name)
         if 'save_to_image_dir' in inference_config:
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/inference/image_inference.py:162]: Good idea!

File: micro_dl/inference/image_inference.py [hunk 23]
```diff
@@ -230,54 +258,48 @@ def __init__(self,
...
         split_col = self.config['dataset']['split_by_column']
         frames_meta = aux_utils.read_meta(self.image_dir)
         inference_ids = np.unique(frames_meta[split_col]).tolist()
         if data_split == 'all':
             return split_col, inference_ids
 
         try:
-            split_fname = os.path.join(self.model_dir, 'split_samples.json')
+            split_dir_name = os.path.join(self.torch_predictor.network_config['model_dir'], os.pardir)
+            split_dir_name = '/'.join(self.torch_predictor.network_config['model_dir'].split('/')[:-1])
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/inference/image_inference.py:301]: Fixed

File: micro_dl/scripts/config_validation_test.py [hunk 24]
```diff
@@ -0,0 +1,15 @@
+from micro_dl.config._torch_config import TorchConfig
+import micro_dl.utils.aux_utils as aux_utils
+import os
+
+kidney_25D_config = os.path.join('/hpc/projects/compmicro/projects',
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/scripts/config_validation_test.py:5]: This was written less as a unit test and more as a functionality check. I didn't intend for this to be included in the final PR. I have been using nose to write some unittests for the network modules. Those can be found under tests.

File: micro_dl/inference/image_inference.py [hunk 25]
```diff
@@ -72,33 +90,35 @@ def __init__(self,
...
-            fnames = natsort.natsorted(fnames)
-            model_fname = fnames[-1]
-
+            raise AssertionError('model_dir must be specified in torch config for inference')
+        
         self.config = train_config
         self.model_dir = model_dir
         self.image_dir = inference_config['image_dir']
 
+        if 'save_folder_name' in inference_config:
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/inference/image_inference.py:109]: I agree, we should do more careful handling of paths. That being said if someone forgets the '/' on essentially any path provided in the config files the code will break, as paths are treated as absolute in a number of other places (I ran into this the first time I ran the code as a user). 

I am relying on users to know that specifying a save folder starting with '/' will make it global, and not doing so will make it local (in this case I've treated local as local to the model dir, rather than the running dir, as I saw Soorya had expected this behavior, so it seemed intuitive to me). Is there another robust way you were thinking of that a user can specify that they want an abs vs. local path used as the save directory?

File: micro_dl/input/dataset.py [hunk 26]
```diff
@@ -305,21 +308,21 @@ def _augment_image(self,
             trans_image = np.rot90(
                 input_image,
                 k=1,
-                axes=(0 + add_dim, 1 + add_dim),
+                axes=(-2, -1),
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/input/dataset.py:311]: Good catch. I changed this because previously it was rotating over channel dimension when input was 4d (as in 2d unet), which caused an error. 

To simplify code, PyTorch implementation only accepts data in format (...zyx or ...zxy). If we want to handle other formats I think the best way is to convert any other formats ((b,z,y,x,c) for example) to (b,c,z,y,x) on read-in and write.

File: micro_dl/scripts/inference_bug_tracing.py [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+import argparse
+import datetime
+import os
+import torch
+import yaml
+import sys
+
+sys.path.insert(0, '/home/christian.foley/virtual_staining/microDL')
+
+import micro_dl.utils.aux_utils as aux_utils
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/scripts/inference_bug_tracing.py:11]: Yep! I've been using the scripts under microDL/micro_dl/scripts to help with development and I don't intend to include any of them in the final PR. If I've written a script that should be included, it will be stored in microDL/scripts.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 28]
```diff
@@ -0,0 +1,226 @@
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from micro_dl.torch_unet.networks.layers.ConvBlock3D import *
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py:6]: Good point! In this case the ConvBlock3D module only contains the ConvBlock3D class, so I thought it was simple enough here, especially because the lines are quite long already.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+        """
+        Instance of 2.5D Unet. 
+        1.) https://elifesciences.org/articles/55502
+        
+        Architecture takes in stack of 2d inputs given as a 3d tensor and returns a 2d interpretation.
+        Learns 3d information based upon input stack, but speeds up training by compressing 3d 
+        information before the decoding path. Uses interruption conv layers in the Unet skip paths to 
+        compress information with z-channel convolution.
+        
+        :param int in_channels: number of feature channels in (phase, retardance)
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py:34]: Yep, I see how this could be confusing.  Fixed.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 30]
```diff
@@ -0,0 +1,226 @@
...
+        :param int out_channels: number of feature channels out (nuclei, f-actin, etc)
+        :param int input_stack_depth: depth of input
+        :param int output_stack_depth: depth of output
+        :param int/tuple(int, int) xy_kernel_size: size of x and y dimensions of conv kernels in blocks
+        :param bool residual: see name
+        :param float dropout: probability of dropout, between 0 and 0.5
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py:40]: Fixed. I was previously relying on PyTorch's dropout function to handle this.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 31]
```diff
@@ -0,0 +1,226 @@
...
+        super(Unet25d, self).__init__()
+        self.in_channels = in_channels
+        self.num_blocks = num_blocks
+        self.ks = xy_kernel_size
+        self.residual = residual
+        self.dropout = dropout
+        self.task = task
+        
+        #----- set static parameters -----#
+        self.block_padding = 'same'
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py]: Yep! Let's discuss what activation functions and pooling types we should give access to as well as the upsampling (from the comment below), how we should allow that access (directly in config vs optional parameters vs inherited from some config class), and what we should select as our defaults.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 32]
```diff
@@ -0,0 +1,309 @@
+import datetime
+import os
+import time
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (jennyfolkesson) [inline micro_dl/torch_unet/networks/Unet25D.py:7]: Are you using all these imports?

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 33]
```diff
@@ -0,0 +1,309 @@
...
+        assert (
+            dropout >= 0 and dropout <= 0.5
+        ), f"Dropout {dropout} not in allowed range: [0, 0.5]"
+        self.dropout = dropout
+        self.task = task
+        self.debug_mode = False
+
+        # ----- set static parameters ----- #
+        self.block_padding = "same"
+        down_mode = "avgpool"  # TODO set static avgpool
```
Comment (jennyfolkesson) [inline micro_dl/torch_unet/networks/Unet25D.py:67]: down_mode and up_mode refers to pooling type and upsampling type, right? Just a gentle reminder that the more clearly you name your variables, the easier it is to read the code.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 34]
```diff
@@ -0,0 +1,309 @@
...
+        forward_filters = [in_channels] + self.num_filters
+        backward_filters = [
+            self.num_filters[-(i + 1)] + self.num_filters[-(i + 2)]
+            for i in range(len(self.num_filters))
+            if i < len(self.num_filters) - 1
+        ] + [out_channels]
+
+        # ----- Downsampling steps -----#
+        self.down_list = []
+        if down_mode == "maxpool":
```
Comment (jennyfolkesson) [inline micro_dl/torch_unet/networks/Unet25D.py:91]: Instead of a number of if statements with for loops embedded in them, it would be cleaner to map valid strings to nn functions. If you'd like to make it easy for yourself you can require your down_mode to match ff function names (e.g. instead of 'maxpool' -> 'MaxPool3d'), in which case you can use getattr directly:
pooling = getattr(nn, down_mode)
for i in range(num_blocks):
    self.down_list.append(pooling(kernel_size=(1, 2, 2)))

That way you can let the user directly choose from torch.nn which type to use. Although you might want to check first that it's within a set that works with the model so things don't break if the user tries to call a 1D pooling or something.

If you would like to keep down_modes like 'maxpool' you can define a dictionary mapping, e.g.:
pool_dict = {'maxpool': 'MaxPool3d', 'avgpool': 'AvgPool3d'}
and then proceed with getattr.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 35]
```diff
@@ -0,0 +1,309 @@
...
+        elif down_mode == "conv":
+            raise NotImplementedError("Not yet implemented!")
+            # TODO: implement.
+        self.register_modules(self.down_list, "down_samp")
+
+        # ----- Upsampling steps ----- #
+        self.up_list = []
+        for i in range(num_blocks):
+            self.up_list.append(
+                nn.Upsample(scale_factor=(1, 2, 2), mode=up_mode, align_corners=False)
```
Comment (jennyfolkesson) [inline micro_dl/torch_unet/networks/Unet25D.py:106]: Did you check that up_mode is a valid mode somewhere?

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 36]
```diff
@@ -0,0 +1,309 @@
...
+        # ----- Standardize Filter Sequence ----- #
+        if len(num_filters) != 0:
+            assert len(num_filters) == num_blocks, (
+                "Length of num_filters must be equal to num_"
+                "blocks + 1 (number of convolutional blocks per path)."
+            )
+            self.num_filters = num_filters
+        else:
+            self.num_filters = [pow(2, i) * 16 for i in range(num_blocks + 1)]
+            self.num_filters
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py:81]: I dont think the row is extra: if I remember correctly I added this during testing to deal with an edge case of a very small Unet with 1 block.

File: micro_dl/torch_unet/networks/Unet25D.py [hunk 37]
```diff
@@ -0,0 +1,309 @@
...
+        forward_filters = [in_channels] + self.num_filters
+        backward_filters = [
```
[the remark this answers is not in the document]
Comment (Christianfoley) [reply micro_dl/torch_unet/networks/Unet25D.py:83]: Great catch! Will change this.

**Comment by mattersoflight** [comment]
Closing this PR, since we have decided to spin the PyTorch implementation as a separate repo soon.
